docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,467 @@
+
+  // effect is the effect for the taint. Valid values are NoSchedule, PreferNoSchedule and NoExecute.
+  // +required
+  // +kubebuilder:validation:Enum=NoSchedule;PreferNoSchedule;NoExecute
+  Effect corev1.TaintEffect `json:"effect"`
+
+  // propagation defines how this taint should be propagated to Nodes.
+  // Always: The taint will be continuously reconciled. If removed from a node, it will be re-added.
+  // OnInitialization: The taint will be added during node initialization. If removed, it will not be re-added.
+  // +required
+  Propagation MachineTaintPropagation `json:"propagation"`
+}
+
+// MachineTaintPropagation defines when a taint should be propagated to Nodes.
+// +kubebuilder:validation:Enum=Always;Initialize
+type MachineTaintPropagation string
+
+const (
+  // TaintPropagationAlways means the taint should be continuously reconciled and kept on the Node.
+  // If removed, it will be re-added during reconciliation.
+  TaintPropagationAlways MachineTaintPropagation = "Always"
+
+  // TaintPropagationOnInitialization means the taint should be set once during initialization and then
+  // left alone. If removed, it will not be re-added.
+  TaintPropagationOnInitialization MachineTaintPropagation = "OnInitialization"
+)
+```
+
+Proper validations on the new field should ensure that no taint with a key of `node.cluster.x-k8s.io/uninitialized` or `node.cluster.x-k8s.io/outdated-revision` is getting added, because that one is restricted to be used by CAPI and providers.
+Other taints normally set by kubeadm should be able to get set.
+
+##### Changes to the Machine, MachineSet, MachineDeployment and MachinePool resources via MachineSpec
+
+Note: To reduce verbosity, this proposal does not include all kinds of required validation markers.
+
+This proposes to add a field array to the `MachineSpec` struct.
+This implicitly leads to adding the field to the following types:
+
+| API Type            | New field                    |
+|---------------------|------------------------------|
+| `Machine`           | `.spec.taints`               |
+| `MachineSet`        | `.spec.template.spec.taints` |
+| `MachineDeployment` | `.spec.template.spec.taints` |
+| `MachinePool`       | `.spec.template.spec.taints` |
+
+```golang
+
+type MachineSpec struct{
+  // taints are the node taints that Cluster API will manage.
+  // This list is not necessarily complete: other Kubernetes components may add or remove other taints.
+  // Only those taints defined in this list will be added or removed by core Cluster API controllers.
+  //
+  // NOTE: This list is implemented as a "map" type, meaning that individual elements can be managed by different owners.
+  // As of Kubernetes 1.33, this is different from the implementation on corev1.NodeSpec, but provides a more flexible API for components building on top of Cluster API.
+  // +optional
+  // +listType=map
+  // +listMapKey=key
+  // +listMapKey=effect
+  Taints []MachineTaint `json:"taints,omitempty"`
```

Comment:
We will want to enforce some bounds on the number of taints, any thoughts on what might be an appropriate answer?

If we capped this to 64 taints for example, do we expect CAPI users to come back and ask us to increase it?

Reply:
Too many taints never make sense because a pod needs to tolerate all taints to get scheduled to a node.

If there are too many taints, it may result in a user to tolerate "all kind of" taints of a given effect.
I think that would be a bad result so limiting totally makes sense 👍 

Afaik we don't limit on labels and annotations, but also the use case is different there I'd say and should not block us setting a limit here.

I'll do some research and come back to this.